### PR TITLE
Optimize the variable declaration, replacing the use of forEach

### PR DESCRIPTION
**Pull request by xguest**
Need not 'Array.every', a 'Array.find', but for 'object' is no such function,  the cycle wrote.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,56 +1,45 @@
 'use strict';
 
-var forEach = require('foreach');
+var toStr = Object.prototype.toString,
```
Comment (ljharb) [inline index.js]: please use one "var" per var - this is an intentional decision so the debugger can step through them.

File: index.js [hunk 2]
```diff
@@ -1,56 +1,45 @@
...
-		if (!descriptor) {
-			var superProto = Object.getPrototypeOf(proto);
-			descriptor = gOPD(superProto, Symbol.toStringTag);
+module.exports = Object.getPrototypeOf && gOPD && typeof Symbol === 'function' && typeof Symbol.toStringTag === 'symbol' ? function (value) {
+	var typedArray, arr, proto, descriptor;
+	/* eslint max-statements: ['error', 16] */
+	/* eslint no-restricted-syntax: ['warn', 'ForInStatement'] */
+	/* jscs:disable disallowNodeTypes */
+	for (typedArray in typedArrays) {
+		if (Object.prototype.hasOwnProperty.call(typedArrays, typedArray)) {
```
Comment (ljharb) [inline index.js]: please use the "has" module instead of a live reference to `Object.prototype.hasOwnProperty`

File: index.js [hunk 3]
```diff
@@ -1,56 +1,45 @@
...
-		var arr = new global[typedArray]();
-		if (!(Symbol.toStringTag in arr)) {
-			throw new EvalError('this engine has support for Symbol.toStringTag, but ' + typedArray + ' does not have the property! Please report this.');
-		}
-		var proto = Object.getPrototypeOf(arr);
-		var descriptor = gOPD(proto, Symbol.toStringTag);
-		if (!descriptor) {
-			var superProto = Object.getPrototypeOf(proto);
-			descriptor = gOPD(superProto, Symbol.toStringTag);
+module.exports = Object.getPrototypeOf && gOPD && typeof Symbol === 'function' && typeof Symbol.toStringTag === 'symbol' ? function (value) {
```
Comment (ljharb) [inline index.js]: please leave these checks in separate variables - the more variables you create rather than inline, the more the JIT can optimize, and the more humans can read it (which is the most important thing).

File: index.js (hunk 2 above)
Comment (xguest) [inline index.js]: Do I understand correctly the module "has", he, in turn, will 'function-bind'? Maybe just create a variable?

Comment (ljharb) [inline index.js]: yes, it will - it preserves a bound copy in the require cache to prevent things like `delete Object.prototype.hasOwnProperty` from breaking this code.

File: .eslintrc [hunk 4]
```diff
@@ -4,6 +4,8 @@
 	"extends": "@ljharb",
 
 	"rules": {
-		"max-statements-per-line": [2, { "max": 2 }]
+		"max-statements-per-line": [2, { "max": 2 }],
+		"max-statements": ["error", 16],
+		"no-restricted-syntax": ["warn", "ForInStatement"]
```
Comment (ljharb) [inline .eslintrc]: for this rule, please leave it at 2, and manually define the entire list with "ForInStatement" removed.

File: .eslintrc [hunk 5]
```diff
@@ -4,6 +4,8 @@
 	"extends": "@ljharb",
 
 	"rules": {
-		"max-statements-per-line": [2, { "max": 2 }]
+		"max-statements-per-line": [2, { "max": 2 }],
+		"max-statements": ["error", 16],
```
Comment (ljharb) [inline .eslintrc]: please use 0/1/2 instead of the string form.

File: .gitignore [hunk 6]
```diff
@@ -1,3 +1,6 @@
+.gitignore
```
Comment (ljharb) [inline .gitignore]: `.gitignore` should absolutely not be ignored, otherwise it can't be committed to the repo.

File: .gitignore [hunk 7]
```diff
@@ -1,3 +1,6 @@
...
+GIT_*.cmd
```
Comment (ljharb) [inline .gitignore]: i'm not familiar with this filetype, but it seems more appropriate in your global gitignore rather than per-project

File: index.js [hunk 8]
```diff
@@ -16,41 +13,44 @@ var typedArrays = {
 	Uint16Array: true,
 	Uint32Array: true
 };
-
 var slice = String.prototype.slice;
-var toStrTags = {};
+var getpr = Object.getPrototypeOf;
```
Comment (ljharb) [inline index.js]: This method is part of ES5 and may not be available in every browser that has Typed Arrays (which shipped before ES5).

File: index.js [hunk 9]
```diff
@@ -16,41 +13,44 @@ var typedArrays = {
...
 var gOPD = Object.getOwnPropertyDescriptor;
-if (hasToStringTag && gOPD && Object.getPrototypeOf) {
-	forEach(typedArrays, function (_, typedArray) {
-		var arr = new global[typedArray]();
-		if (!(Symbol.toStringTag in arr)) {
-			throw new EvalError('this engine has support for Symbol.toStringTag, but ' + typedArray + ' does not have the property! Please report this.');
+var isTypedArray;
+if (getpr && gOPD) {
+	if (typeof Symbol === 'function') {
+		if (typeof Symbol.toStringTag === 'symbol') {
```
Comment (ljharb) [inline index.js]: let's de-nest this by restoring the `var hasToStringTag = typeof Symbol === 'function' && typeof Symbol.toStringTag === 'symbol';` line

File: index.js [hunk 10]
```diff
@@ -16,41 +13,44 @@ var typedArrays = {
...
-		if (!anyTrue) {
-			try {
-				anyTrue = getter.call(value) === typedArray;
-			} catch (e) { /**/ }
-		}
-	});
-	return anyTrue;
-};
-
-module.exports = function isTypedArray(value) {
```
Comment (ljharb) [inline index.js]: I intentionally return the same function in all cases - since the purpose of this PR is the "for-each" optimization, please revert all unrelated changes.

File: index.js [hunk 11]
```diff
@@ -18,34 +18,31 @@ var typedArrays = {
...
-}
-
 var tryTypedArrays = function tryTypedArrays(value) {
-	var anyTrue = false;
-	forEach(toStrTags, function (getter, typedArray) {
-		if (!anyTrue) {
-			try {
-				anyTrue = getter.call(value) === typedArray;
-			} catch (e) { /**/ }
+	if (Object.getPrototypeOf) {
```
Comment (ljharb) [inline index.js]: this should be determined at the module level, so that it can't be faked later.

File: index.js [hunk 12]
```diff
@@ -18,34 +18,31 @@ var typedArrays = {
...
+				if (descriptor) {
+					return typedArrays[descriptor.get.call(Object(value))];
+				}
+			}
 		}
-	});
-	return anyTrue;
+	}
+	/* jscs:enable disallowNodeTypes */
+	return false;
```
Comment (ljharb) [inline index.js]: wouldn't this mean that in a browser without `Object.getPrototypeOf`, this will always return false? This module needs to work in pre-ES5 browsers that have TypedArrays.

Comment (xguest) [inline index.js]: I told you about this
P.S. My opinion: it is an indirect dependency.
Offer: instead of false
typedArrays [slice.call (toStr.call (value), 8, -1)];

File: index.js (hunk 11 above)
Comment (xguest) [inline index.js]: Where do you intend to add a check, you have it in the function.

Comment (xguest) [inline index.js]: Sorry not in the function.

Comment (xguest) [inline index.js]: Please look at my first option. Poor English, I find it hard to understand.

File: index.js (hunk 12 above)
Comment (ljharb) [inline index.js]: It is a dependency of your implementation - it is not a dependency of mine. Thus, any "improvement" must not add this indirect dependency.

Certainly your "instead of false" suggestion might work - but this PR is about replacing the for-each. Any unrelated changes should be saved for other PRs.

**Comment by xguest** [comment]
I'm sorry, line 23 of your implementation, check 'if (hasToStringTag && gOPD && Object.getPrototypeOf) {' can not create an object that ignores the function tryTypedArrays.

I have a question
Why:
I intentionally return the same function in all cases - since the purpose of this PR is the "for-each" optimization, please revert all unrelated changes.

Please explain why not use:
 
```javascript
 value instanceof arr // by constructor
 arr.prototype.isPrototypeOf (value) // by prototype

```
Thank you for your hard work

**Comment by ljharb** [comment]
`instanceof` and `isPrototypeOf` and `.constructor` all don't work if the value comes from, say, an iframe, or a web worker.

**Comment by xguest** [comment]
I have reviewed your code, it gets the name of choice on the type of 'Symbol.toStringTag' or 'Object.prototype.toString'.
All checks are necessary to carry out 'Symbol.toStringTag'.
I really do not know much JavaScript, I think you need to add more options for obtaining type name, or select one unique.
It was nice to talk to, I would like to continue the dialogue.
The purest version of your code that I managed to get:
```JavaScript
'use strict';

var typedArrays = /.*((Float|Uint|Int)(8Clamped|8|16|32|64))Array.*/,
	def = function (value) {
		var out,
			val = Object.prototype.toString.call(value);
		return (out = val.replace(typedArrays, '$1')) !== val && out;
	};

module.exports = Object.getPrototypeOf && Object.getOwnPropertyDescriptor && typeof Symbol === 'function' && typeof Symbol.toStringTag === 'symbol' ? function (value) {
	var proto, descriptor, out, val;
	/* eslint no-return-assign: 1 */
	return Object(value) === value && (
		Symbol.toStringTag in value ? (
				descriptor = Object.getOwnPropertyDescriptor(
					proto = Object.getPrototypeOf(value),
					Symbol.toStringTag
				) || Object.getOwnPropertyDescriptor(
					Object.getPrototypeOf(proto),
					Symbol.toStringTag
				)
			) && descriptor.get && (
				val = descriptor.get.call(value),
				out = val.replace(typedArrays, '$1')
			) !== val && out : def(value)
  );
} : def;

```

**Comment by ljharb** [comment]
Object property lookups after module parse time are brittle - anyone could overwrite `Object.getPrototypeOf` or `Object.getOwnPropertyDescriptor` and break the code. All methods needed should be looked up beforehand, and stored in variables accessible via closure to the exported function.

Some other suggestions are - avoid trying to condense everything to a single expression, and avoid trying to reduce variable usage. More variables are better, both for readability and optimizability. In addition, multiline ternaries should always be avoided.

**Comment by xguest** [comment]
What if:
```javascript
module.exports = function () {
	// Perform preliminary checks and return the appropriate function.
	return function () {/* exactly what is needed */}
}();
```

**Comment by ljharb** [comment]
There's no need to do that in the IIFE, but sure, that would work - except then you have a different function depending on the environment, which is harder to test, and negligibly faster.

**Comment by xguest** [comment]
I tried to take into account all the remarks:
```JavaScript
'use strict';

var toStr = Object.prototype.toString;
var hasToStringTag = typeof Symbol === 'function' && typeof Symbol.toStringTag === 'symbol';

var typedArrays = /.*((Float|Uint|Int)(8Clamped|8|16|32|64))Array.*/;
var gOPD = Object.getOwnPropertyDescriptor;
var gProt = Object.getPrototypeOf;

module.exports = function isTypedArray(value) {
	/* eslint max-statements: ['error', 16] */
	if (Object(value) === value) {
		var proto, superProto, descriptor, val, out;
		if (gProt && gOPD && hasToStringTag) {
			if (Symbol.toStringTag in value) {
				proto = gProt(value);
				descriptor = gOPD(proto, Symbol.toStringTag);
				if (!descriptor) {
					superProto = gProt(proto);
					descriptor = gOPD(superProto, Symbol.toStringTag);
				}
				if (descriptor && descriptor.get) {
					val = descriptor.get.call(value);
				}
			}
		} else {
			val = toStr.call(value);
		}
		if (val) {
			out = val.replace(typedArrays, '$1');
			return out !== val && out;
		}
	}
	return false;
};
```

File: index.js [hunk 13]
```diff
@@ -1,6 +1,6 @@
 'use strict';
 
-var forEach = require('foreach');
+var getprototypeof = require('getprototypeof');
```
Comment (ljharb) [inline index.js]: unfortunately this module does not actually fall back to `Object.getPrototypeOf` when available - and it also relies on `constructor`, which is mutable.

File: index.js [hunk 14]
```diff
@@ -18,39 +18,27 @@ var typedArrays = {
...
-	return anyTrue;
-};
 
 module.exports = function isTypedArray(value) {
-	if (!value || typeof value !== 'object') { return false; }
-	if (!hasToStringTag) { return !!typedArrays[slice.call(toStr.call(value), 8, -1)]; }
-	if (!gOPD) { return false; }
-	return tryTypedArrays(value);
+	var out;
+	if (Object(value) === value) {
```
Comment (ljharb) [inline index.js]: if this condition is false, then it's a primitive, and it can never be a typed array - the early return false here is preferred (which is what line 52 originally handled)

**Comment by xguest** [comment]
Hi!
My hold Object.getPrototypeOf?

**Comment by ljharb** [comment]
This PR title is "replace the use of the forEach" - please restrict the PR to just that change.

It's very hard to review, and understand, a large PR that changes more than one thing. I appreciate the contributions, but I can't safely accept them if they are too risky, and right now this one is too risky.

File: index.js [hunk 15]
```diff
@@ -18,34 +16,25 @@ var typedArrays = {
 };
 
 var slice = String.prototype.slice;
-var toStrTags = {};
 var gOPD = Object.getOwnPropertyDescriptor;
-if (hasToStringTag && gOPD && Object.getPrototypeOf) {
-	forEach(typedArrays, function (_, typedArray) {
-		var arr = new global[typedArray]();
-		if (!(Symbol.toStringTag in arr)) {
-			throw new EvalError('this engine has support for Symbol.toStringTag, but ' + typedArray + ' does not have the property! Please report this.');
```
Comment (ljharb) [inline index.js]: I'd still like to keep this error, so I can be alerted if an engine adds the Symbol without adding it to all the right places.

File: index.js [hunk 16]
```diff
@@ -18,34 +16,25 @@ var typedArrays = {
...
-var tryTypedArrays = function tryTypedArrays(value) {
-	var anyTrue = false;
-	forEach(toStrTags, function (getter, typedArray) {
-		if (!anyTrue) {
-			try {
-				anyTrue = getter.call(value) === typedArray;
-			} catch (e) { /**/ }
+var tryTypedArrays = function isTypedArray(value) {
+	var typedArray;
+	if (Object.getPrototypeOf) {
```
Comment (ljharb) [inline index.js]: since these are now evaluated at runtime, this is not robust against `delete Object.getPrototypeOf`, or `global.Symbol = {};`. Any value used at runtime must be evaluated and cached at module parse time.

File: index.js [hunk 17]
```diff
@@ -20,31 +20,37 @@ var typedArrays = {
...
 var toStrTags = {};
 var gOPD = Object.getOwnPropertyDescriptor;
+var typedArray;
 if (hasToStringTag && gOPD && Object.getPrototypeOf) {
-	forEach(typedArrays, function (_, typedArray) {
-		var arr = new global[typedArray]();
-		if (!(Symbol.toStringTag in arr)) {
-			throw new EvalError('this engine has support for Symbol.toStringTag, but ' + typedArray + ' does not have the property! Please report this.');
+	/* jscs:disable disallowNodeTypes */
+	for (typedArray in typedArrays) {
```
Comment (ljharb) [inline index.js]: How is this an improvement over the `forEach` call? Loops should, in general, be avoided.

File: index.js [hunk 18]
```diff
@@ -20,31 +20,37 @@ var typedArrays = {
...
 	var anyTrue = false;
-	forEach(toStrTags, function (getter, typedArray) {
-		if (!anyTrue) {
-			try {
-				anyTrue = getter.call(value) === typedArray;
-			} catch (e) { /**/ }
+	/* jscs:disable disallowNodeTypes */
+	for (typedArray in toStrTags) {
+		if (has(toStrTags, typedArray)) {
+			anyTrue = toStrTags[typedArray].call(value);
```
Comment (ljharb) [inline index.js]: i think if this is doing an early return, and since "anyTrue" isn't a good variable name for it any more, maybe it should do `if (toStrTags[typedArray].call(value) === typedArray) { return true; }`?

**Comment by xguest** [comment]
Hi!
I removed check descriptor.get in function

As far as I understand you.
If I change forEach on for this good change that has no meaning.
If we rewrite the function without cycles is a bad thing:
- Dangerous Object.getPrototypeOf
- Not monitored typedArrays. "Symbol.toStringTag in arr"
- It is difficult to understand the code - a lot of changes.

Tell where and what I can do to you profit and I could then what  learn?

**Comment by xguest** [comment]
You said that before the module.exports a safe space and I can use open functions or constructors.
For example:
toStrTags [typedArray] = arr.constructor;

It is safe?
P.S. This does not apply to the PR

**Comment by ljharb** [comment]
The "constructor" property isn't safe in any use case.

**Comment by xguest** [comment]
Hi.
Unfortunately, not enough time, work has appeared.
Thank you for your hard work.
It was nice to chat.
Bye.
